java/adapter/jdbc/src/main/java/org/apache/arrow/adapter/jdbc/JdbcToArrowUtils.java

```diff
@@ -188,12 +188,9 @@ public static ArrowType getArrowTypeFromJdbcType(final JdbcFieldInfo fieldInfo,
       case Types.TIME:
         return new ArrowType.Time(TimeUnit.MILLISECOND, 32);
       case Types.TIMESTAMP:
-        final String timezone;
-        if (calendar != null) {
-          timezone = calendar.getTimeZone().getID();
-        } else {
-          timezone = null;
-        }
+        return new ArrowType.Timestamp(TimeUnit.MILLISECOND);
+      case Types.TIMESTAMP_WITH_TIMEZONE:
+        final String timezone = calendar == null ? null : calendar.getTimeZone().getID();
```

Comment:
Hmm, there is still an underlying timezone, right? I guess we just don't know without writing database-specific code. It might be better to error in this case?

Reply:
Yeah this one is tough. As I understand it, `TIMESTAMP_WITH_TIMEZONE` means that each record includes its own timezone as a UTC offset (or at least, that seems to be the case with Snowflake). Based on the JDBC spec, `ResultSet.getTimestamp(int columnIndex, Calendar cal)` does the following `This method uses the given calendar to construct an appropriate millisecond value for the timestamp if the underlying database does not store timezone information.`. This leads me to believe that it parses out the offset from the underlying db and uses that to convert to UTC. If no offset is available, it assumes the timestamp is in w/e is passed in the `calendar` and then converts it to UTC. Here's an interesting SO on the topic: https://stackoverflow.com/a/63078938/1815486

From my understanding, `ArrowType.Timestamp` does not support per record timezones and it does not do any conversions on its own. It seems like the timezone is associated with the Vector itself and all records are expected to be in that same timezone. This means we must convert `TIMESTAMP_WITH_TIMEZONE` values into one specific TZ before we add them to the vector. It's also not exactly clear what it means when the TZ is `null` (I'm assuming this means it's just a "wall clock" time).

This all leaves some questions:
- What is the TZ associated with `calendar` supposed to represent for these functions? 
  - The TZ to assume the underlying values are in
  - The TZ to report on the ArrowType
  - The TZ that we should convert values to
  
There's some important connotations because it depends on whether one is using the arrow-jdbc library to convert values server side, where it's generally safe to use/assume UTC for a lot of things, or client side, where they might want to express values in the user's timezone.

Reply:
In Arrow, the underlying representation for a timestamp with timezone is always UTC. So if the driver is giving us UTC and converting it, we should always bypass their conversion and construct a `timestamp[ms, UTC]`.

And yes, if there's no timezone, then it's a wall-clock time.

Reply:
Hmmm maybe we can change the consumer to do something smarter than `getTimestamp`. [This SO](https://stackoverflow.com/a/73092871/1815486) suggests `OffsetDateTime odt = myResultSet.getObject( … , OffsetDateTime.class ) ; `.

So for TZ aware timestamps we could use ^ and if no offset is available assume UTC. I think that would work for most things, but it'd get tricky for something like `TIMESTAMP_LTZ`. I wonder how that even comes across through the JDBC driver? I imagine it's still of type `TIMESTAMP_WITH_TIMEZONE` just like `TIMESTAMP_TZ`. We might want to add the value from `rsmd.getColumnTypeName` to the `JdbcFieldInfo` as well so users can disambiguate between the two if they need to write a custom converter.

Reply:
I ran into the same issue: https://github.com/apache/arrow/issues/35916 

So agreed, we need more info than what JdbcFieldInfo has. (ADBC's Java adapter already has a workaround to add the extra fields.) And yeah, I think trying to get OffsetDateTime or similar would be better.

That's part of why I want to vendor this into the ADBC driver, iterate on it + test it against actual databases, and then maybe send the changes back...

Reply:
Cool makes sense. I'll try to modify the consumer to do something smarter and correctly convert values into UTC.

I'll make a separate PR to add more data into `JdbcFieldInfo`.

WRT `I want to vendor this into the ADBC driver... and then maybe send the changes back` you mean redesign the JDBC to Arrow conversion within ADBC and eventually pull that back out into the `arrow-jdbc` package?

Reply:
Yeah. Though realistically, I don't have the time to do that right now so thank you for fixing up these things :sweat_smile: